From f4d2c8439f5753e06b2414872262d4c4f673137a Mon Sep 17 00:00:00 2001
From: Score_Under <seejay.11@gmail.com>
Date: Wed, 3 Oct 2012 03:30:38 +0100
Subject: [PATCH] Fix clientside sign blanking If a block break on a sign or
 mob spawner was cancelled, the client would display an
 empty sign or a pig spawner in its place. This fixes that
 issue by re-sending any additional tile-entity data
 whenever it re-sends a block that is not known to not be a
 tile entity.

---
 .../net/minecraft/server/ItemInWorldManager.java   |   12 ++++++------
 .../net/minecraft/server/NetServerHandler.java     |   12 ++++++------
 .../org/bukkit/craftbukkit/entity/CraftPlayer.java |   11 +++++++++++
 3 files changed, 23 insertions(+), 12 deletions(-)

diff --git a/src/main/java/net/minecraft/server/ItemInWorldManager.java b/src/main/java/net/minecraft/server/ItemInWorldManager.java
index 025c25a..e83373c 100644
--- a/src/main/java/net/minecraft/server/ItemInWorldManager.java
+++ b/src/main/java/net/minecraft/server/ItemInWorldManager.java
@@ -117,7 +117,7 @@ public class ItemInWorldManager {
             // CraftBukkit start
             if (event.isCancelled()) {
                 // Let the client know the block still exists
-                ((EntityPlayer) this.player).netServerHandler.sendPacket(new Packet53BlockChange(i, j, k, this.world));
+                this.player.getBukkitEntity().refreshBlock(i, j, k, this.world);
                 return;
             }
             // CraftBukkit end
@@ -156,7 +156,7 @@ public class ItemInWorldManager {
                 if (event.useItemInHand() == Event.Result.DENY) {
                     // If we 'insta destroyed' then the client needs to be informed.
                     if (toolDamage > 1.0f) {
-                        ((EntityPlayer) this.player).netServerHandler.sendPacket(new Packet53BlockChange(i, j, k, this.world));
+                        this.player.getBukkitEntity().refreshBlock(i, j, k, this.world);
                     }
                     return;
                 }
@@ -164,7 +164,7 @@ public class ItemInWorldManager {

                 if (blockEvent.isCancelled()) {
                     // Let the client know the block still exists
-                    ((EntityPlayer) this.player).netServerHandler.sendPacket(new Packet53BlockChange(i, j, k, this.world));
+                    this.player.getBukkitEntity().refreshBlock(i, j, k, this.world);
                     return;
                 }

@@ -214,7 +214,7 @@ public class ItemInWorldManager {
             }
         // CraftBukkit start - force blockreset to client
         } else {
-            ((EntityPlayer) this.player).netServerHandler.sendPacket(new Packet53BlockChange(i, j, k, this.world));
+            this.player.getBukkitEntity().refreshBlock(i, j, k, this.world);
             // CraftBukkit end
         }
     }
@@ -279,7 +279,7 @@ public class ItemInWorldManager {

             if (event.isCancelled()) {
                 // Let the client know the block still exists
-                ((EntityPlayer) this.player).netServerHandler.sendPacket(new Packet53BlockChange(i, j, k, this.world));
+                this.player.getBukkitEntity().refreshBlock(i, j, k, this.world);
                 return false;
             }
         }
@@ -301,7 +301,7 @@ public class ItemInWorldManager {
             boolean flag = this.d(i, j, k);

             if (this.isCreative()) {
-                this.player.netServerHandler.sendPacket(new Packet53BlockChange(i, j, k, this.world));
+                this.player.getBukkitEntity().refreshBlock(i, j, k, this.world); // CraftBukkit - refresh the block correctly
             } else {
                 ItemStack itemstack = this.player.bP();
                 boolean flag1 = this.player.b(Block.byId[l]);
diff --git a/src/main/java/net/minecraft/server/NetServerHandler.java b/src/main/java/net/minecraft/server/NetServerHandler.java
index 3374639..1301383 100644
--- a/src/main/java/net/minecraft/server/NetServerHandler.java
+++ b/src/main/java/net/minecraft/server/NetServerHandler.java
@@ -543,20 +543,20 @@ public class NetServerHandler extends NetHandler {
                 // CraftBukkit start
                 if (i1 < this.server.getSpawnRadius() && !flag) {
                     CraftEventFactory.callPlayerInteractEvent(this.player, Action.LEFT_CLICK_BLOCK, i, j, k, l, this.player.inventory.getItemInHand());
+                    this.player.getBukkitEntity().refreshBlock(i, j, k, worldserver);
                     // CraftBukkit end
-                    this.player.netServerHandler.sendPacket(new Packet53BlockChange(i, j, k, worldserver));
                 } else {
                     this.player.itemInWorldManager.dig(i, j, k, packet14blockdig.face);
                 }
             } else if (packet14blockdig.e == 2) {
                 this.player.itemInWorldManager.a(i, j, k);
                 if (worldserver.getTypeId(i, j, k) != 0) {
-                    this.player.netServerHandler.sendPacket(new Packet53BlockChange(i, j, k, worldserver));
+                    this.player.getBukkitEntity().refreshBlock(i, j, k, worldserver); // CraftBukkit - refresh block correctly
                 }
             } else if (packet14blockdig.e == 1) {
                 this.player.itemInWorldManager.c(i, j, k);
                 if (worldserver.getTypeId(i, j, k) != 0) {
-                    this.player.netServerHandler.sendPacket(new Packet53BlockChange(i, j, k, worldserver));
+                    this.player.getBukkitEntity().refreshBlock(i, j, k, worldserver); // CraftBukkit - refresh block correctly
                 }
             } else if (packet14blockdig.e == 3) {
                 double d4 = this.player.locX - ((double) i + 0.5D);
@@ -565,7 +565,7 @@ public class NetServerHandler extends NetHandler {
                 double d7 = d4 * d4 + d5 * d5 + d6 * d6;

                 if (d7 < 256.0D) {
-                    this.player.netServerHandler.sendPacket(new Packet53BlockChange(i, j, k, worldserver));
+                    this.player.getBukkitEntity().refreshBlock(i, j, k, worldserver); // CraftBukkit - refresh block correctly
                 }
             }
         }
@@ -654,7 +654,7 @@ public class NetServerHandler extends NetHandler {
         }

         if (flag) {
-            this.player.netServerHandler.sendPacket(new Packet53BlockChange(i, j, k, worldserver));
+            this.player.getBukkitEntity().refreshBlock(i, j, k, worldserver); // CraftBukkit - refresh block correctly
             if (l == 0) {
                 --j;
             }
@@ -679,7 +679,7 @@ public class NetServerHandler extends NetHandler {
                 ++i;
             }

-            this.player.netServerHandler.sendPacket(new Packet53BlockChange(i, j, k, worldserver));
+            this.player.getBukkitEntity().refreshBlock(i, j, k, worldserver); // CraftBukkit - refresh block correctly
         }

         itemstack = this.player.inventory.getItemInHand();
diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
index a1e9f85..17da361 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
@@ -281,6 +281,17 @@ public class CraftPlayer extends CraftHumanEntity implements Player {
         playEffect(loc, effect, datavalue);
     }

+    public void refreshBlock(int x, int y, int z, net.minecraft.server.World world) {
+        this.getHandle().netServerHandler.sendPacket(new Packet53BlockChange(x, y, z, world));
+        TileEntity tileEntity = world.getTileEntity(x, y, z);
+        if (tileEntity != null) {
+            Packet updateTileEntity = tileEntity.l();
+            if (updateTileEntity != null) {
+                this.getHandle().netServerHandler.sendPacket(updateTileEntity);
+            }
+        }
+    }
+
     public void sendBlockChange(Location loc, Material material, byte data) {
         sendBlockChange(loc, material.getId(), data);
     }
--
1.7.4.4
